lib/ruby_lsp/requests/hover.rb

```diff
@@ -0,0 +1,99 @@
+    # it renders a clickable link to that DSL's official documentation at https://api.rubyonrails.org
+    #
+    # # Example
+    #
+    # ```ruby
+    # class Post < ApplicationRecord
+    #   before_save :do_something # when hovering on before_save, the link will be rendered
+    # end
+    # ```
+    class Hover < BaseRequest
+      extend T::Sig
+
+      sig { params(document: Document, uri: String, position: Document::PositionShape).void }
+      def initialize(document, uri, position)
+        super(document)
+
+        @file_dir = T.let(Pathname.new(uri).dirname.to_s, String)
+        @hover = T.let(nil, T.nilable(LanguageServer::Protocol::Interface::Hover))
+        position = Document::Scanner.new(document.source).find_position(position)
+        @target = T.let(find(T.must(@document.tree), position), T.nilable(SyntaxTree::Node))
+      end
+
+      sig { override.returns(T.nilable(LanguageServer::Protocol::Interface::Hover)) }
+      def run
+        return if @target.nil?
+
+        visit(@document.tree)
+        @hover
+      end
+
+      sig { override.params(node: SyntaxTree::Command).void }
+      def visit_command(node)
+        return unless node == @target
+
+        message = node.message
+        generate_rails_document_link_hover(message.value, message)
+      end
+
+      sig { override.params(node: SyntaxTree::ConstPathRef).void }
+      def visit_const_path_ref(node)
+        return unless node == @target
+
+        constant_name = full_constant_name(node)
+        generate_rails_document_link_hover(constant_name, node)
+      end
+
+      private
+
+      sig { params(name: String, node: SyntaxTree::Node).void }
+      def generate_rails_document_link_hover(name, node)
+        urls = Support::RailsDocumentClient.generate_rails_document_urls(name, @file_dir)
+
+        unless urls.empty?
+          values = urls.map do |url|
+            "[Rails Document: `#{url[:name]}`](#{url[:url]})"
+          end
+          contents = LanguageServer::Protocol::Interface::MarkupContent.new(
+            kind: "markdown",
+            value: values.join("\n"),
+          )
+          @hover = LanguageServer::Protocol::Interface::Hover.new(
+            range: range_from_syntax_tree_node(node),
+            contents: contents,
+          )
+        end
+      end
+
+      sig do
+        params(node: SyntaxTree::Node, position: Integer).returns(T.nilable(SyntaxTree::Node))
+      end
+      def find(node, position)
+        matched = node.child_nodes.compact.bsearch do |child|
+          if (child.location.start_char...child.location.end_char).cover?(position)
+            0
+          else
+            position <=> child.location.start_char
+          end
+        end
+
+        case matched
+        when SyntaxTree::Command, SyntaxTree::ConstPathRef
+          matched
+        when SyntaxTree::Node
+          find(matched, position)
+        end
+      end
```

Comment:
Yeah big fan of doing this in syntax tree. I'd be super amenable to adding this. Only thing I would change is that this should say `while matched in SyntaxTree::Node` instead of recursing.

Reply:
@kddnewton I know that you want to keep the Syntax Tree light and thus didn't have a position to node map in place, but I wonder if clients could ask to have such a map when calling the `parse` method and Syntax Tree could start keeping track of such a thing to make operations like this to be constant time?